Converter, folders, tests

marl0vv opened the pull request.

marl0vv committed:
Add ConverterService

- Added converter that takes initial image in inputStream and returns image converted to webp in a byte array

marl0vv committed:
ConverterControllers works with ConverterService

- Stub was deleted and now controller returns converted image

marl0vv committed:
Add Service tests

marl0vv committed:
Add Converter Controller Tests

RavenVSS commented on src/main/java/com/glamik/webpconverter/controllers/ConverterController.java:
```diff
@@ -0,0 +1,33 @@
+
+    private final ConverterService converterService;
+
+    public ConverterController(ConverterService converterService) {
```
Почитай про lombok, добавь его в проект  

Он умеет генерировать конструкторы сам, и ещё кучу всего упрощает

marl0vv replied:
Добавил. Теперь конструктор генерируется через lombok

RavenVSS commented on pom.xml:
```diff
@@ -1,54 +1,61 @@
+        <dependency>
+            <groupId>org.sejda.webp-imageio</groupId>
+            <artifactId>webp-imageio-sejda</artifactId>
+            <version>0.1.0</version>
```
Судя по репозиторию, есть версия повыше, но название артефакта чуть отличается, может быть это и не оно - посмотри

https://mvnrepository.com/artifact/org.sejda.imageio/webp-imageio

marl0vv replied:
Исправил на новую версию. Это оно, всё хорошо. На старую наткнулся через поисковик и сразу добавил. Видимо, они в какой-то момент решили поменять название артефакта или что-то.

RavenVSS commented on src/main/java/com/glamik/webpconverter/service/ConverterService.java:
```diff
@@ -0,0 +1,21 @@
+        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
+            BufferedImage image = ImageIO.read(inputStream);
+            ImageIO.write(image, "webp", outputStream);
+            return outputStream.toByteArray();
```
Попробуй поработать с временными файлами, т.е. в сервис передавать не InputStream а File, также и возвращать

Есть удобная либа для этого
https://github.com/antkorwin/ioutils

marl0vv replied:
Поменял на временные файлы. 
Библиотеку подключить не смог. При добавлении её в проект, у меня не запускается программа, даже если я никакие функции из этой библиотеки не использую. 

Код ошибки:
java: java.lang.ClassCastException: class org.jetbrains.jps.javac.$Proxy15 cannot be cast to class com.sun.tools.javac.processing.JavacProcessingEnvironment (org.jetbrains.jps.javac.$Proxy15 is in unnamed module of loader java.net.URLClassLoader @38af3868; com.sun.tools.javac.processing.JavacProcessingEnvironment is in module jdk.compiler of loader 'app')

RavenVSS commented on src/test/java/com/glamik/webpconverter/service/ConverterServiceTests.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Test
+    void testSuccessfulConversion() throws IOException {
+        InputStream inputStream = Files.newInputStream(Path.of("src/test/resources/test-image.jpg"));
```
Есть способы чуть поудобнее получать файлы из ресурсов
https://www.baeldung.com/spring-classpath-file-access

marl0vv replied:
Теперь файлы из ресурсов получаю вот так:

`InputStream convertedStream = new ClassPathResource("/test-image.webp").getInputStream();`

RavenVSS commented on src/test/java/com/glamik/webpconverter/service/ConverterServiceTests.java:
```diff
@@ -0,0 +1,56 @@
+        InputStream inputStream = Files.newInputStream(Path.of("src/test/resources/test-image.jpg"));
+        byte[] webpBytes = converterService.convertToWebp(inputStream);
+
+        assertThat(webpBytes).isNotNull().isNotEmpty();
```
`isNotEmpty()` скорей всего включает проверку на null, можно один метод убрать

RavenVSS commented on src/test/java/com/glamik/webpconverter/service/ConverterServiceTests.java:
```diff
@@ -0,0 +1,56 @@
+        assertThat(webpBytes).isNotNull().isNotEmpty();
+
+        BufferedImage image = ImageIO.read(new ByteArrayInputStream(webpBytes));
+        assertThat(image).isNotNull();
```
Тут лучше сравнить с эталонным значением. В ресурсы помести правильный webp и потом с ним сравнивай

RavenVSS commented on src/test/java/com/glamik/webpconverter/controllers/ConverterControllerTests.java:
```diff
@@ -0,0 +1,68 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@SpringBootTest
```
Эта аннотация поднимает контекст спринга (сервис целиком)  
Нужно сделать unit тесты, без подъёма контекста - сами тесты нормальные, но надо поднятие контекста убрать

Объекты тестирования создаёшь как поля класса
Вот пример таких тестов
https://github.com/antkorwin/xsync/blob/master/src/test/java/com/antkorwin/xsync/MultiKeysXSyncTest.java

marl0vv replied:
Немного почитал про подъём контекста и аннотации.
Использовал @WebMvcTest(ConverterController.class). Пойдёт или стоит вообще убрать все аннотации?

marl0vv committed:
Changed webp-imageio to a newer version

marl0vv committed:
Add lombok

- Added lombok
- Changed constructor in controller to lombok RequiredArgsConstructor annotation

marl0vv committed:
ConverterService changed to use Path class instead of streams

marl0vv committed:
Change ConverterController to work with files

Now ConverterController makes temp folder and creates temp input image in it and temp output image. Then output image goes as a return to request and temp files are cleaned.

marl0vv committed:
Converter and Service tests changed

RavenVSS commented on src/main/java/com/glamik/webpconverter/controllers/ConverterController.java:
```diff
@@ -0,0 +1,64 @@
+    public ResponseEntity<PathResource> convertImage(@RequestParam("image") MultipartFile imageFile) {
+        Path tempDir;
+        try {
+            tempDir = Files.createTempDirectory("webp-converter-");
```
Без временной дериктории

RavenVSS commented on src/main/java/com/glamik/webpconverter/controllers/ConverterController.java:
```diff
@@ -0,0 +1,64 @@
+
+    private String getFileExtension(String filename) {
+        int lastDot = filename.lastIndexOf('.');
+        return (lastDot == -1) ? ".tmp" : filename.substring(lastDot);
```
Может быть такое, что расширение не указано, но узнать тип нужно

Почитай про https://www.baeldung.com/apache-tika  
Нужно получить mimeType файла и уже по нему определять - можем мы это обработать или нет

RavenVSS replied:
Вроде как там можно и чистое название файла получить - без расширения, надо посмотреть

RavenVSS replied:
Это оставь, но вынеси в сервис проверку mimeType

RavenVSS commented on src/main/java/com/glamik/webpconverter/service/ConverterService.java:
```diff
@@ -0,0 +1,34 @@
+@Service
+public class ConverterService {
+
+    public Path convertToWebp(Path inputPath) throws IOException {
```
В этот метод надо добавить проверку mimeType, что мы можем такое обработать (см. коммент выше)

RavenVSS replied:
В нём же можно и определять тип файла, т.е. вынести из контроллера часть логики

RavenVSS commented on src/test/java/com/glamik/webpconverter/controllers/ConverterControllerTests.java:
```diff
@@ -0,0 +1,71 @@
+
+
+@WebMvcTest(ConverterController.class)
+class ConverterControllerTests {
```
Тут всё равно контекст поднимается  
Создай контроллер через `ConverterController controller = new ConverterController(service)`, а сервис создай через `Mockito.mock()`

RavenVSS commented on src/main/java/com/glamik/webpconverter/controllers/ConverterController.java:
```diff
@@ -0,0 +1,64 @@
+            webpPath = converterService.convertToWebp(tempInputPath);
+            PathResource resource = new PathResource(webpPath);
+
+            return ResponseEntity.ok().body(resource);
```
Добавить хедер Content-Disposition

RavenVSS commented:
Для тестов добавь:
```
// Arrange
        
// Act

// Assert
```

marl0vv committed:
Delete temp dir

marl0vv committed:
Changed Path class to File class

marl0vv committed:
Add MimeType check

Check MIME-type of input file. If not image then throws IllegalArgumentException.
Exception is getting caught in controller and returns 400 bad request

marl0vv committed:
Service and Controller Tests fixes

- Changed controller testx in a way that they are not using application context anymore
- Added Arrange-Act-Assert comments to Service and Controller tests

marl0vv committed:
Added responseHeaders

RavenVSS commented on src/main/java/com/glamik/webpconverter/service/ConverterService.java:
```diff
@@ -1,34 +1,40 @@
-        Path outputPath = inputPath.getParent().resolve(outputFileName);
+    public File convertToWebp(File inputFile) throws IOException {
+        checkInputMimeType(inputFile);
+        File outputFile = new File(inputFile.getParent() + "/ConvertedTempImage.webp");
```
Можно тоже через tempFile

RavenVSS commented on src/main/java/com/glamik/webpconverter/service/ConverterService.java:
```diff
@@ -1,34 +1,40 @@
-        int lastDot = filename.lastIndexOf('.');
-        return (lastDot == -1) ? filename : filename.substring(0, lastDot);
+    private void checkInputMimeType(File inputFile) throws IOException {
+        Tika tika = new Tika();
```
Вынести в static поле класса

RavenVSS commented on src/test/java/com/glamik/webpconverter/controllers/ConverterControllerTests.java:
```diff
@@ -2,70 +2,78 @@
-
-        when(converterService.convertToWebp(any(Path.class))).thenReturn(samplePath);
-
+        // Arrange
```
Добавить
```
when(converterService.convertToWebp(any(File.class))).thenReturn(*returnFile*)
```

RavenVSS commented on src/test/java/com/glamik/webpconverter/controllers/ConverterControllerTests.java:
```diff
@@ -2,70 +2,78 @@
 
-        Files.deleteIfExists(samplePath);
+        // Assert
+        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
```
Проверить body и headers

RavenVSS commented on src/test/java/com/glamik/webpconverter/service/ConverterServiceTests.java:
```diff
@@ -1,63 +1,70 @@
         byte[] referenceBytes = referenceStream.readAllBytes();
+
+        // Assert
         assertThat(convertedBytes).isEqualTo(referenceBytes);
```
Сравнивать outputFile с эталонным

marl0vv committed:
Service and Controller Tests fixes
